### PR TITLE
Add Pangolin to nextstrain-base docker image

### LINKED ISSUE #33: Add Pangolin tools
## Context
[We recently added support for reannotating Pangolin lineages as part of the ncov workflow](https://github.com/nextstrain/ncov/pull/593/files). To support these annotations, we needed to install `pangolin=2.3.8` and `pango-learn=2021.04.01` in the Snakemake Conda environment.

In a previous Nextstrain meeting, we discussed whether to include Pangolin tools in this base image or not given that these tools only work for the ncov workflow and are not needed for any other workflow that uses the "base" image. The consensus at the time was to install Pangolin tools in the base image to make maintenance easier and avoid supporting yet another image that is specific to the ncov workflow.

## Description

To support Pangolin annotations from the ncov workflow, we need to [install pangolin and pango-learn](https://cov-lineages.org/resources/pangolin/installation.html). The creators of Pangolin recommend installation with Conda, but we do not currently support Conda in the Docker image, so we need to install Pangolin and its dependencies manually based on the contents of [the most recent Conda environment file](https://github.com/cov-lineages/pangolin/blob/master/environment.yml).

## Testing

If the software is installed properly, we should be able to run an ncov workflow with `run_pangolin: true` in a `builds.yaml` configuration.

### PR DESCRIPTION
**Pull request by jgadling**
### Description of proposed changes    
This adds Pangolin to nextstrain-base docker image. I'm not sure specifically which version of pangolin we want to include, so I used the versions mentioned in the GitHub issue. Just let me know if I should update this!

### Related issue(s)  

Fixes #33 

### Testing
I've rebuilt the docker image and Pangolin runs successfully:

```
% ./devel/build
<output omitted>
% docker run -ti --rm=true nextstrain/base -- sh -c 'pangolin /usr/local/lib/python3.7/site-packages/pangolin/data/reference.fasta 2> /dev/null && echo output: && cat lineage_report.csv'
Found the snakefile
Do things with sam files

Usage:
  gofasta sam [flags]
  gofasta sam [command]

Available Commands:
  indels       Parse a SAM file for raw indel information
  toMultiAlign Convert a SAM file to a multiple alignment in fasta format
  toPairAlign  convert a SAM file to pairwise alignments in fasta format
  variants     Find mutations relative to a reference from an alignment in sam format

Flags:
  -h, --help               help for sam
  -r, --reference string   Reference fasta file used to generate the sam file
  -s, --samfile string     Samfile to read. If none is specified, will read from stdin (default "stdin")
  -t, --threads int        Number of threads to use (default 1)

Use "gofasta sam [command] --help" for more information about a command.
2.24-r1122
5.10.0
The query file is:/usr/local/lib/python3.7/site-packages/pangolin/data/reference.fasta
Looking in /usr/local/lib/python3.7/site-packages/pangoLEARN/data for data files...

Data files found
Trained model:  /usr/local/lib/python3.7/site-packages/pangoLEARN/data/decisionTree_v1.joblib
Header file:    /usr/local/lib/python3.7/site-packages/pangoLEARN/data/decisionTreeHeaders_v1.joblib
Lineages csv:   /usr/local/lib/python3.7/site-packages/pangoLEARN/data/lineages.metadata.csv
loading model 01/11/2022, 02:07:00
processing block of 1 sequences 01/11/2022, 02:07:01
complete 01/11/2022, 02:07:02
Output file written to: /nextstrain/build/lineage_report.csv
output:
taxon,lineage,probability,pangoLEARN_version,status,note
outgroup_A,A,1.0,2021-04-01,passed_qc,
```

### CODE REVIEW

**Review by tsibley — CHANGES_REQUESTED** [review]
Thanks for this PR, @jgadling!

This Dockerfile uses a [two-stage build](https://docs.docker.com/develop/develop-images/multistage-build/) with most of the work happening in the first stage `builder` image. The second stage then copies artifacts from the builder image over into the final image, keeping the final image free of things only needed for building (e.g. git, a C toolchain, etc).

The code added by this PR is in the second-stage final image but should instead be in the first-stage `builder` image. Could you rework the PR to move it? You'll also then not need the added `git` dep in the second stage.

File: Dockerfile [hunk 1]
```diff
@@ -151,6 +152,15 @@ COPY --from=builder \
...
 COPY --from=builder /build/vcftools/built/share/  /usr/local/share/
 
+# Add Pangolin and PangoLEARN + deps
+RUN curl -fsSL https://github.com/virus-evolution/gofasta/releases/download/v0.0.6/gofasta-linux-amd64 \
+        -o /usr/local/bin/gofasta \
+  && chmod a+rx /usr/local/bin/gofasta
+RUN cd /usr/local/bin && curl -fsSL https://github.com/lh3/minimap2/releases/download/v2.24/minimap2-2.24_x64-linux.tar.bz2 \
+  | tar xjvpf - --strip-components=1 minimap2-2.24_x64-linux/minimap2
+RUN pip install git+https://github.com/cov-lineages/pangolin.git@v2.3.8
+RUN pip install git+https://github.com/cov-lineages/pangoLEARN.git@2021-04-01
```
Comment (huddlej) [inline Dockerfile:162]: The pango versions mentioned in the original issue that this PR addresses are pretty old now. I think we'll want the latest versions of `pangolin` ([3.1.17](https://github.com/cov-lineages/pangolin/releases/tag/v3.1.17)) and pangoLEARN ([2021-12-06](https://github.com/cov-lineages/pangoLEARN/releases/tag/2021-12-06)).

It looks like [pangolin version 3.0+ requires UShER](https://cov-lineages.org/resources/pangolin/installation.html), too. UShER provides [an installation script for Ubuntu](https://github.com/yatisht/usher/blob/master/install/installUbuntu.sh) that we could use as a starting point.

**Review by huddlej — COMMENTED** [review]
Thank you for putting this together, @jgadling and @sidneymbell! My only additional comment to Tom's is about trying this out with the latest pango versions. It seems we'll need to install UShER for the latest pangolin, but let us know if this is more work than you can commit to at the moment. We can pick up from this PR.

File: Dockerfile (hunk 1 above)
Comment (jgadling) [reply Dockerfile]: Ok - I've managed to get Pangolin 3.x with UShER functionality working in the image build, but if I'm reading the Pango docs correctly, UShER is an *optional* dependency for pangolin ([docs](https://cov-lineages.org/resources/pangolin/usage.html)) and I don't see anything in nexstrain that enables that functionality, so I'm omitting it.

**Comment by sidneymbell** [comment]
Hey there @tsibley and @huddlej -- is this branch now good to merge? :)

File: Dockerfile [hunk 2]
```diff
@@ -172,10 +185,16 @@ COPY --from=builder /usr/local/lib/python3.7/site-packages/ /usr/local/lib/pytho
 # troublesome or excessive.
 #   -trs, 15 June 2018
 COPY --from=builder \
+    /usr/local/bin/*.py \
```
Comment (huddlej) [inline Dockerfile:188]: I tried running pango annotations from the ncov workflow with this line commented out and everything worked as expected.

Are there specific Python scripts we should include here? Even if there are a lot, an explicit list would help with future maintenance.

**Review by huddlej — COMMENTED** [review]
Thank you for the updates, @jgadling! You were absolutely right about not needing UShER for our ncov use case; it's nice to avoid that extra complexity in this file.

I tested this Dockerfile like so:

```
# Pull and build.
./devel/pull
./devel/build

# Run pango annotations in ncov workflow with Nextstrain CI build.
cd ../ncov
nextstrain build --cpus 2 --image nextstrain/base:latest . -np --forceall --configfile nextstrain_profiles/nextstrain-ci/builds.yaml --config run_pangolin=True
```

This all worked mostly as expected, although I got a couple of scikit-learn warnings that I hadn't noticed in our ncov conda environment before. These warnings seem to stem from using a later version of scikit-learn in our Docker image that was used to prepare the pango model files:

```
/usr/local/lib/python3.7/site-packages/sklearn/base.py:338: UserWarning: Trying to unpickle estimator DecisionTreeClassifier from version 0.24.2 when using version 1.0.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/modules/model_persistence.html#security-maintainability-limitations
  UserWarning,
/usr/local/lib/python3.7/site-packages/sklearn/base.py:444: UserWarning: X has feature names, but DecisionTreeClassifier was fitted without feature names
  f"X has feature names, but {self.__class__.__name__} was fitted without"
```

I visually confirmed that the pango annotations looked correct, though, and there were no errors. I haven't dug into why we have scikit-learn installed, but these warnings don't seem associated with a real problem.

I made a note below about this, but I didn't see a specific need to copy any additional Python scripts from `/usr/local/bin` in the second stage of the image build. Were there specific scripts you found were required in your own testing? If not, we can drop that line and merge the rest of the changes as they are.

File: Dockerfile (hunk 2 above)
Comment (jgadling) [reply Dockerfile]: This could have been left over from the attempt to install usher, you're right that I don't think it's necessary. I've removed it.

**Review by huddlej — APPROVED** [review]
Looks good to me, @jgadling! Thank you again for your work on this issue!

**Pull request merged**